File: examples/python/Depth/unified_depth_user_cameras.py
#!/usr/bin/env python3
"""
Same flow as unified_depth.py, but left/right ``Camera`` nodes are created explicitly on the device's
first stereo pair.

For stereo backends (``AUTO`` -> ``StereoDepth`` / ``NeuralDepth`` / …, or explicit non-ToF algorithms),
``Depth`` reuses those cameras and requests stereo-sized outputs (it does not adopt the camera nodes).

On RVC2 with a ToF sensor, ``AUTO`` resolves to the ToF backend: the pre-built stereo cameras are not used
for depth. On RVC4 the NeuralDepth zoo model inside ``Depth`` is fixed (not user-configurable).

Create ``depth_node.depth`` / ``confidence`` queues **before** ``pipeline.build()``; wiring happens on first access.

This script still requires ``getStereoPairs()`` to be non-empty so the demo can build left/right cameras (they are unused when ``AUTO`` selects ToF on RVC2).
"""

import sys

import cv2
import depthai as dai
import numpy as np

pipeline = dai.Pipeline()
device = pipeline.getDefaultDevice()
if device is None:
    print("Connect a device (host-only pipeline cannot use Depth).", file=sys.stderr)
    sys.exit(1)

stereoPairs = device.getStereoPairs()
if not stereoPairs:
    print("This device has no stereo pair; Depth cannot run.", file=sys.stderr)
    sys.exit(1)

stereoPair = stereoPairs[0]

pipeline.create(dai.node.Camera).build(stereoPair.left, sensorResolution=(1280, 800), sensorFps=30)
pipeline.create(dai.node.Camera).build(stereoPair.right, sensorResolution=(1280, 800), sensorFps=30)

depthNode = pipeline.create(dai.node.Depth)

depthQueue = depthNode.depth.createOutputQueue()
confidenceQueue = depthNode.confidence.createOutputQueue()

pipeline.build()

colorMap = cv2.applyColorMap(np.arange(256, dtype=np.uint8), cv2.COLORMAP_JET)
colorMap[0] = [0, 0, 0]


def colorizeDepthMm(frame: np.ndarray) -> np.ndarray:
    valid = frame > 0
    if not np.any(valid):
        return np.zeros((*frame.shape, 3), dtype=np.uint8)
    vmax = float(np.max(frame[valid]))
    norm = np.zeros_like(frame, dtype=np.uint8)
    norm[valid] = ((frame[valid].astype(np.float32) / vmax) * 255).astype(np.uint8)
    return cv2.applyColorMap(norm, colorMap)


def colorizeConfidence(frame: np.ndarray) -> np.ndarray:
    if frame.dtype == np.uint16:
        vmax = int(np.max(frame))
        if vmax <= 0:
            return np.zeros((*frame.shape, 3), dtype=np.uint8)
        vis = ((frame.astype(np.float32) / vmax) * 255).astype(np.uint8)
    else:
        vis = frame
    return cv2.applyColorMap(vis, colorMap)


with pipeline:
    pipeline.start()
    while pipeline.isRunning():
        depthFrame = depthQueue.get()
        assert isinstance(depthFrame, dai.ImgFrame)
        cv2.imshow("depth (user cameras)", colorizeDepthMm(depthFrame.getFrame()))

        confidenceFrame = confidenceQueue.get()
        assert isinstance(confidenceFrame, dai.ImgFrame)
        cv2.imshow("confidence", colorizeConfidence(confidenceFrame.getFrame()))

        key = cv2.waitKey(1)
        if key == ord("q"):
            pipeline.stop()
            break

    cv2.destroyAllWindows()

File: examples/cpp/Depth/unified_depth_user_cameras.cpp
#include <opencv2/opencv.hpp>

#include <iostream>

#include "depthai/depthai.hpp"

namespace {

cv::Mat colorizeDepthMm(const cv::Mat& frame) {
    if(frame.empty() || frame.channels() != 1) {
        return cv::Mat::zeros(frame.size(), CV_8UC3);
    }

    const cv::Mat validMask = frame > 0;
    if(cv::countNonZero(validMask) == 0) {
        return cv::Mat::zeros(frame.size(), CV_8UC3);
    }

    double maxValue = 0.0;
    cv::minMaxLoc(frame, nullptr, &maxValue, nullptr, nullptr, validMask);

    cv::Mat normalized(frame.size(), CV_8U, cv::Scalar(0));
    if(maxValue > 0.0) {
        cv::Mat scaled;
        frame.convertTo(scaled, CV_32F, 255.0 / maxValue);
        scaled.convertTo(normalized, CV_8U);
    }

    cv::Mat colorized;
    cv::applyColorMap(normalized, colorized, cv::COLORMAP_JET);
    colorized.setTo(cv::Scalar::all(0), ~validMask);
    return colorized;
}

cv::Mat colorizeConfidence(const cv::Mat& frame) {
    if(frame.empty() || frame.channels() != 1) {
        return cv::Mat::zeros(frame.size(), CV_8UC3);
    }

    cv::Mat normalized;
    if(frame.depth() == CV_16U) {
        double maxValue = 0.0;
        cv::minMaxLoc(frame, nullptr, &maxValue);
        if(maxValue <= 0.0) {
            return cv::Mat::zeros(frame.size(), CV_8UC3);
        }
        cv::Mat scaled;
        frame.convertTo(scaled, CV_32F, 255.0 / maxValue);
        scaled.convertTo(normalized, CV_8U);
    } else {
        frame.convertTo(normalized, CV_8U);
    }

    cv::Mat colorized;
    cv::applyColorMap(normalized, colorized, cv::COLORMAP_JET);
    return colorized;
}

}  // namespace

int main() {
    dai::Pipeline pipeline;
    auto device = pipeline.getDefaultDevice();
    if(device == nullptr) {
        std::cerr << "Connect a device (host-only pipeline cannot use Depth)." << std::endl;
        return 1;
    }

    const auto stereoPairs = device->getStereoPairs();
    if(stereoPairs.empty()) {
        std::cerr << "This device has no stereo pair; Depth cannot run." << std::endl;
        return 1;
    }

    const auto stereoPair = stereoPairs[0];

    pipeline.create<dai::node::Camera>()->build(stereoPair.left, std::make_pair(1280u, 800u), 30.0f);
    pipeline.create<dai::node::Camera>()->build(stereoPair.right, std::make_pair(1280u, 800u), 30.0f);

    auto depthNode = pipeline.create<dai::node::Depth>();

    auto depthQueue = depthNode->depth().createOutputQueue();
    auto confidenceQueue = depthNode->confidence().createOutputQueue();

    pipeline.build();
    pipeline.start();

    while(pipeline.isRunning()) {
        auto depthFrame = depthQueue->get<dai::ImgFrame>();
        auto confidenceFrame = confidenceQueue->get<dai::ImgFrame>();

        if(depthFrame != nullptr) {
            cv::imshow("depth (user cameras)", colorizeDepthMm(depthFrame->getFrame()));
        }
        if(confidenceFrame != nullptr) {
            cv::imshow("confidence", colorizeConfidence(confidenceFrame->getFrame()));
        }

        const int key = cv::waitKey(1);
        if(key == 'q') {
            pipeline.stop();
            break;
        }
    }

    cv::destroyAllWindows();
    return 0;
}
